### PR TITLE
Proper remote host checking in S3 redirects

### PR DESCRIPTION
**Pull request by excitoon**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Proper remote host checking in S3 redirects (security-related thing).

### CODE REVIEW

File: src/IO/S3Common.h [hunk 1]
```diff
@@ -22,6 +23,17 @@ namespace DB
 namespace DB::S3
 {
 
+struct ExtendedClientConfiguration : public Aws::Client::ClientConfiguration
```
[the remark this answers is not in the document]
Comment (excitoon) [reply src/IO/S3Common.h]: +

File: src/IO/S3Common.h [hunk 2]
```diff
@@ -22,6 +23,17 @@ namespace DB
...
+{
+    const Context & context;
```
[the remark this answers is not in the document]
Comment (excitoon) [reply src/IO/S3Common.h]: +

**Comment by excitoon** [comment]
I've just found out that proxy resolver is not being checked for host filter as well. :-(

**Review by Akazz — APPROVED** [review]
LGTM

File: src/IO/S3/PocoHTTPClientFactory.cpp [hunk 3]
```diff
@@ -11,7 +11,7 @@ namespace DB::S3
 std::shared_ptr<Aws::Http::HttpClient>
 PocoHTTPClientFactory::CreateHttpClient(const Aws::Client::ClientConfiguration & clientConfiguration) const
 {
-    return std::make_shared<PocoHTTPClient>(clientConfiguration);
+    return std::make_shared<PocoHTTPClient>(static_cast<const PocoHTTPClientConfiguration &>(clientConfiguration));
```
Comment (alexey-milovidov) [inline src/IO/S3/PocoHTTPClientFactory.cpp:14]: Here is the bug.

**Pull request merged**